bpo-29636: Add --indent / --no-indent arguments to json.tool

dhimmel opened the pull request:
From https://github.com/python/cpython/pull/201 which is being split into two pull requests.

Tagging @serhiy-storchaka, @berkerpeksag, @methane who helped review #201.

Closes http://bugs.python.org/issue29636.

<!-- issue-number: bpo-29636 -->
https://bugs.python.org/issue29636
<!-- /issue-number -->

dhimmel committed:
Add --indent / --no-indent arguments to json.tool

From https://github.com/python/cpython/pull/201 which is being split into
two pull requests.

See http://bugs.python.org/issue29636

dhimmel commented:
Note to rebase on https://github.com/python/cpython/pull/346 which should be merged first.

vstinner commented on Lib/json/tool.py:
```diff
@@ -25,24 +25,35 @@ def main():
+    group.add_argument('--indent', default=4, type=int,
+                       help='Indent level (number of spaces) for '
+                       'pretty-printing. Defaults to 4.')
+    group.add_argument('--no-indent', action='store_const', dest='indent',
```
I would prefer a --compact option which use indent=None but also separators without spaces: http://bugs.python.org/issue29540

It's not the same than what you propose.

dhimmel replied:
@haypo I agree the most compact representation would be useful.

One option would be add a third argument (`--compact`) to the mutually exclusive group. This would provide access to three options:

1. `--indent`: pretty (with newlines)
2. `--no-indent`: with `indent=None` in `json.dump`
3. `--compact`: compact, with `separators=(',', ':')` in `json.dump`

Alternatively, are you proposing to only support options 1 & 3?

I'm happy to implement this. Let's get a second opinion, so I don't go down the wrong track. Tagging @serhiy-storchaka, @berkerpeksag, @methane.

vstinner replied:
Is --no-indent differen than --indent=0?

Would it be possible to use a default of 4 spaces when using --indent with no argument?

serhiy-storchaka replied:
> Is --no-indent differen than --indent=0?

Yes, it is. `--indent=0` adds new lines, `--no-indent` writes all in one line.

> Would it be possible to use a default of 4 spaces when using --indent with no argument?

This is the default behavior.

dhimmel replied:
@haypo the following illustrates the three options

```python
import json
obj = [1, 2]

print('--indent=4')
print(json.dumps(obj, indent=4))

print('--no-indent')
print(json.dumps(obj, indent=None))

print('--compact')
print(json.dumps(obj, separators=(',', ':')))
```

which produces the following output

```
--indent=4
[
    1,
    2
]
--no-indent
[1, 2]
--compact
[1,2]
```

dhimmel replied:
> Spaces after comma and semicolon are not just for readability. They make JSON compatible with YAML 1.0.

@serhiy-storchaka great point. I think this is a strong argument for not replacing `--no-indent` with `--compact`. However, should we add `--compact` as a third option? I can see the utility for users who wan't the smallest file sizes.

serhiy-storchaka commented:
Spaces after comma and semicolon are not just for readability. They make JSON compatible with YAML 1.0.

vstinner commented:
> Yes, it is. --indent=0 adds new lines, --no-indent writes all in one line.

"write all in one line": it sounds like the use case is to produce
smaller file, so it seems redundant and less efficient than --compact.

I expect that --compact removes all spaces and newlines to produce the
smallest file.

serhiy-storchaka commented:
This is not the right place for the design discussion.  Lets continue it on the bug tracker.

dhimmel commented:
> This is not the right place for the design discussion. Lets continue it on the bug tracker.

@serhiy-storchaka and @haypo I updated [bpo-29636](http://bugs.python.org/issue29636#msg288918) with a summary of the design discussion thus far.

I propose continuing this pull request with only `--indent` and `--no-indent`. If we end up reaching consensus to add `--compact`, that can be added in the future. I think it would fit in nicely alongside `--indent` and `--no-indent`. In other words, it's not pressing to decide on `--compact` right now.

dhimmel commented:
https://github.com/python/cpython/pull/346 has been merged and this PR rebased.

@serhiy-storchaka, @berkerpeksag, @methane and @haypo: open for review. No more outstanding changes on my end.

dhimmel committed:
json.tool: --tab identation option

dhimmel committed:
json.tool: --compact identation option

dhimmel committed:
bpo-30971: Improve code readability of json.tool

Originall from https://github.com/python/cpython/pull/2720

Set default option for infile and outfile as per
https://github.com/python/cpython/pull/2720#pullrequestreview-50224377

Use --sort-keys help message based on the json.tool docs at
https://docs.python.org/3.6/library/json.html#basic-usage:

Remove commens as per https://bugs.python.org/msg298692.
Code was descriptive without the comments.

methane committed:
Merge branch 'master' into indent

dhimmel committed:
Add NEWS entry

Use `blurb add` command to assist with adding the NEWS entry.

dhimmel committed:
Update whitespace option documentation

serhiy-storchaka commented on Lib/json/tool.py:
```diff
@@ -21,23 +21,42 @@ def main():
                    'to validate and pretty-print JSON objects.')
     parser = argparse.ArgumentParser(prog=prog, description=description)
     parser.add_argument('infile', nargs='?', type=argparse.FileType(),
+                        default=sys.stdin,
```
What is the benefit of setting default to `sys.stdin` instead of keeping it `None`?

dhimmel replied:
This way a suggestion from another contributor at https://github.com/python/cpython/pull/2720#discussion_r127620912:

> it is easier to read when we process all options in one place, and then use the process options.

By adding `default=sys.stdin`, we remove the following line later on:

```python
infile = options.infile or sys.stdin
```

serhiy-storchaka commented on Lib/json/tool.py:
```diff
@@ -21,23 +21,42 @@ def main():
-    parser.add_argument('--sort-keys', action='store_true', default=False,
-                        help='sort the output of dictionaries alphabetically by key')
+    parser.add_argument('--sort-keys', action='store_true',
+                        help='sort the output of dictionaries by key')
```
Why removed "alphabetically"?

dhimmel replied:
I believe my intention was to make the help message more closely mirror the `json.dump` documentation:

https://github.com/python/cpython/blob/cfa797c0681b7fef47cf93955fd06b54ddd09a7f/Lib/json/__init__.py#L157-L158

Happy to re-add "alphabetically" if that was helpful.

dhimmel committed:
Merge master & resolve --json-lines conflicts
